### PR TITLE
Feat/t218 bls sig arkworks

### PR DESCRIPTION
**Pull request by philippecamacho**
Closes: #218

### CODE REVIEW

File: primitives/src/signatures/bls_over_bn254.rs [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    }
+
+    /// Get the internal of the signing key, namely a P::ScalarField element
+    pub fn sign_key_internal(&self) -> &ScalarField {
+        &self.sk.0
+    }
+
+    /// Signature function
+    pub fn sign<B: AsRef<[u8]>>(&self, msg: &[u8], csid: B) -> Signature {
+        let msg_input = [msg, csid.as_ref()].concat();
```
Comment (chancharles92) [inline primitives/src/signatures/bls_over_bn254.rs:272]: Do we need to hash the message into a fixed-length digest first?

**Review by chancharles92 — COMMENTED** [review]
Thanks @philippecamacho ! Left a few comments.

File: primitives/src/signatures/bls_over_bn254.rs [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        Ok(())
+    }
+
+    /// Sample a pair of keys.
+    fn key_gen<R: CryptoRng + RngCore>(
+        _pp: &Self::PublicParameter,
+        prng: &mut R,
+    ) -> Result<(Self::SigningKey, Self::VerificationKey), PrimitivesError> {
+        let kp = KeyPair::generate(prng);
+        Ok((kp.sk.clone(), kp.vk.clone()))
```
[the remark this answers is not in the document]
Comment (philippecamacho) [reply primitives/src/signatures/bls_over_bn254.rs]: Yes, it is possible to remove this call to `clone`. See 18f5243b.

[the remark this answers is not in the document]
Comment (philippecamacho) [reply primitives/src/signatures/bls_over_bn254.rs]: Yes, it is possible to remove this call to `clone`. See 18f5243b.

[the remark this answers is not in the document]
Comment (chancharles92) [reply primitives/src/signatures/bls_over_bn254.rs]: Ah I was wondering if it's OK to move `kp.vk` without cloning/copying. Anyway, it's a minor comment.

File: primitives/src/signatures/bls_over_bn254.rs [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+/// y^2=x^3+3)   3. Check if `Y` is a quadratic residue (QR), in which case
+/// return `y=sqrt(Y)` otherwise try with `x+1, x+2` etc... until `Y` is a QR.
+///   4. Return `P=(x,y)`
+///
+///  In the future we may switch to a constant time algorithm such as Fouque-Tibouchi <https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf>
+/// * `H` - parameterizable hash function (e.g. SHA256, Keccak)
+/// * `msg` - input message
+/// * `returns` - A group element in G1
+#[allow(non_snake_case)]
+pub fn hash_to_curve<H: Default + DynDigest + Clone>(msg: &[u8]) -> G1Projective {
```
[the remark this answers is not in the document]
Comment (philippecamacho) [reply primitives/src/signatures/bls_over_bn254.rs:214]: The `hash_to_field` function calls [self.expander.expand](https://github.com/arkworks-rs/algebra/blob/bc991d44c5e579025b7ed56df3d30267a7b9acac/ff/src/fields/field_hashers/expander/mod.rs#L68) where most of the logic is placed. In particular the message is first hashed with a bytes to bytes hash function such as SHA256 of Keccak and then some field(s) elements are derived. Does this answer your question?

File: primitives/src/signatures/bls_over_bn254.rs (hunk 1 above)
Comment (philippecamacho) [reply primitives/src/signatures/bls_over_bn254.rs:272]: As discussed [above](https://github.com/EspressoSystems/jellyfish/pull/223#discussion_r1158635887), I don' t think this is necessary.

File: primitives/src/signatures/bls_over_bn254.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chancharles92) [reply primitives/src/signatures/bls_over_bn254.rs:214]: OK, so essentially the `hash_to_field` already somewhat executes the hash-to-digest step internally. Make sense, thanks!

File: primitives/src/signatures/bls_over_bn254.rs [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+                let pk = key_pair.ver_key_ref();
+                assert!(pk.verify(&msg, &sig, CS_ID_BLS_BN254).is_ok());
+                // wrong public key
+                assert!(pk_bad.verify(&msg, &sig, CS_ID_BLS_BN254).is_err());
+                // wrong message
+                msg.push(i as u8);
+                assert!(pk.verify(&msg, &sig, CS_ID_BLS_BN254).is_err());
+            }
+        }
+    }
```
Comment (chancharles92) [inline primitives/src/signatures/bls_over_bn254.rs:355]: Shall we additionally test 2 messages m1, m2=(m1||x) whose lengths are larger than a base field length? E.g., |m1| = 35B, |x| = 1B

Comment (philippecamacho) [reply primitives/src/signatures/bls_over_bn254.rs:355]: Yes it is a good idea, but why do we need two messages? Is it not enough to test with m1?

Comment (philippecamacho) [reply primitives/src/signatures/bls_over_bn254.rs:355]: E.g:

```rust
// Test for long messages
let size=35; // Bigger than 32 which is the number of bytes needed to encode a field element
let key_pair = KeyPair::generate(&mut rng);
let mut msg = [33u8;35];
let sig = key_pair.sign(&msg, CS_ID_BLS_BN254);
let pk = key_pair.ver_key_ref();
assert!(pk.verify(&msg, &sig, CS_ID_BLS_BN254).is_ok());
```

Comment (chancharles92) [reply primitives/src/signatures/bls_over_bn254.rs:355]: To check that a signature for m1 won't verify for m2, or vice versa.

Comment (chancharles92) [reply primitives/src/signatures/bls_over_bn254.rs:355]: The above code looks good. Maybe setup a `msg2 = [33u8; 36]` and add one more line at the bottom:
`assert!(pk.verify(&msg2, &sig, CS_ID_BLS_BN254).is_err());`

Comment (philippecamacho) [reply primitives/src/signatures/bls_over_bn254.rs:366]: Ok, done in a07c2293.

**Review by chancharles92 — APPROVED** [review]

**Pull request merged**